### DISCUSSION TITLE
Move authentication logic to Next.js 12 middleware

### DISCUSSION THREAD

**Discussion by mgm1313** [supabase/supabase#3680] [Questions] [answered]
Now that [Next.js 12 introduces middleware](https://nextjs.org/blog/next-12#introducing-middleware) it would be cool if we could move the authentication workflow to this middleware layer. 

**Current situation**
At this time when requiring server-side authentication the workflow would be to set up an auth cookie via for example an api route. 

[`pages/api/auth.js`](https://github.com/vercel/next.js/blob/46ddd8cf6016bd74b98a6f627149c95fa59244ed/examples/with-supabase-auth-realtime-db/pages/api/auth.js)
```js 
import { supabase } from '../../lib/initSupabase'

export default function handler(req, res) {
  supabase.auth.api.setAuthCookie(req, res)
}
```

which get's called inside the `onAuthStateChanged` method. 

Now to create a protected route, we currently check if the user object obtained via `getUserByCookie` is not empty inside the `getServerSideProps` method of that page. 

[`pages/protected.js`](https://github.com/vercel/next.js/blob/46ddd8cf6016bd74b98a6f627149c95fa59244ed/examples/with-supabase-auth-realtime-db/pages/profile.js#L31)
```js
export async function getServerSideProps({ req }) {
  const { user } = await supabase.auth.api.getUserByCookie(req)

  if (!user) {
    // If no user, redirect to index.
    return { props: {}, redirect: { destination: '/', permanent: false } }
  }

  // If there is a user, return it.
  return { props: { user } }
}
```

**Advantages to using middleware**
When using the new middleware functionality in Next.js 12 I see three main advantages for my own use-case:
1. No more repetitive code; currently I need to copy-paste the `getServerSideProps` method into all page files that are protected
2. Less risk on forgetting to protect a route; e.q. to protect all routes inside the `/dashboard` folder I'd simply have to setup a single `_middleware.js` file inside the `/pages/dashboard` folder containing the authentication logic. 
3. Possibly faster page loads; since the authentication logic runs at the edge instead of on the server. 

**How to get this to work?**
I tried setting up a middleware layer through the following code:

`_middleware.ts`
```ts
import type { NextRequest } from 'next/server'

import { supabase } from '@/utils/initSupabase'

function jsonResponse(status: number, data: any, init?: ResponseInit) {
  return new Response(JSON.stringify(data), {
    ...init,
    status,
    headers: {
      ...init?.headers,
      'Content-Type': 'application/json',
    },
  })
}

async function verifyAuth(request: NextRequest) {
  const { user, error } = await supabase.auth.api.getUserByCookie(request)

  if (!user) {
    return jsonResponse(401, { error: { message: error?.message } })
  }

  return
}

export async function middleware(req: NextRequest) {
  return await verifyAuth(req)
}
```

But this leads to an `XMLHttpRequest is not defined` error inside the `getUser()` method, I guess because those kinds of requests can't be made inside the V8 engine that edge functions are using. 
So is it actually currently possible to use Next.js middleware to authenticate users via Supabase, to achieve above use-case?

**Comment by mgm1313** [comment, marked as answer]
So I dug a little more into this; and had quite the guidance from the [jwt-authentication](https://github.com/vercel/examples/tree/main/edge-functions/jwt-authentication) Vercel example. 

So since I'm not interested in the actual user object, I can just verify the JWT token using my own JWT-secret (which can be obtained from the Supabase dashboard under 'Settings' > 'API' > 'Config' > 'JWT Secret' ). 
To verify the token in the V8 runtime ([which edge functions use](https://nextjs.org/docs/api-reference/edge-runtime)) you can for example use the [`@tsndr/cloudflare-worker-jwt`](https://www.npmjs.com/package/@tsndr/cloudflare-worker-jwt) package, as Vercel does in her example. 

So I ended up with the following `_middleware.ts` file inside my `pages/dashboard` folder. 
```ts
import jwt from '@tsndr/cloudflare-worker-jwt'
import { NextRequest, NextResponse } from 'next/server'

/**
 * Verifies the user's JWT token and returns the payload if
 * it's valid or a response if it's not.
 */
export async function middleware(request: NextRequest) {
  const token = request.cookies['sb:token'] //'sb:token' is the default cookie name

  if (!token || !(await jwt.verify(token, process.env.SUPABASE_JWT_SECRET!))) {
    return NextResponse.redirect('/', 302) // If a user is not authenticated (either no token was send, or the token is invalid) redirect the user to the homepage where they will be presented with a log-in screen
  }

  return NextResponse.next() // continue the middleware chain https://nextjs.org/docs/api-reference/next/server#nextresponse 
}
```

Now I was able to remove all `getServerSideProps` methods throughout my app, and achieve the advantages I [mentioned earlier](https://github.com/supabase/supabase/discussions/3680#discussion-3652380) 🥳

**Reply by mzaien** [reply]
Your solution is very nice 🤟, you can add 
Return NextResponse.next() if the token is available

got the idea from this tweet https://twitter.com/balazsorban44/status/1453482637625110539?s=20

**Reply by balazsorban44** [reply]
Is `sb:token` an `HttpOnly` cookie? In that case, how do you show the user state on the client, since I don't see the token value can be passed down to the page

referring to https://twitter.com/vvoyer/status/1453658292803276804?t=KYhnGSd8JdKlgHwKmXEtPQ&s=19

**Reply by i-pip** [reply]
@mgm1313 this is really lit 🔥  I'd like to send some swag your way for this, can I send you a 100% discount code for the [Supabase store](https://store.supabase.com) on discord - I've got a few discount codes to share 😋

**Reply by matepapp** [reply]
am I the only one getting this error message? 🤔 

```
{ [NotSupportedError: Unrecognized key import format "pkcs8".]
  stack: 'Error: Unrecognized key import format "pkcs8".\n    at Object.verify (worker.js:181:2908)\n    at Object.h [as handler] (worker.js:181:17747)\n    at j (worker.js:181:47835)\n    at Object.g (worker.js:181:3556)\n    at worker.js:205:3671\n    at worker.js:205:986',
  code: 9,
  name: 'NotSupportedError',
  message: 'Unrecognized key import format "pkcs8".' }
```

I'm not a cryptography expert but I'm assuming it should be something related to the encryption format, right?

**Comment by balazsorban44** [comment]
```ts
import { supabase } from '../../lib/initSupabase'

export default function handler(req, res) {
  supabase.auth.api.setAuthCookie(req, res)
}
```

This could just be:

```ts
import { supabase } from '../../lib/initSupabase'

export default supabase.auth.api.setAuthCookie
```

**Comment by fergusmeiklejohn** [comment]
Supabase can't (mustn't) privilege one JS framework over another, never mind one React framework over all the others. Supabase needs to be 100% front end agnostic.

**Reply by danestves** [reply]
We are not talking about that, we are searching for a way to use Supabase in Service Workers because at the end of the day, the middleware of NextJS is a service worker
